### PR TITLE
feat: [IDP-489]: Bitbucket on-prem support for integrations

### PR DESCRIPTION
**Pull request by devesh-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47800)

### CODE REVIEW

**Comment by devesh-harness** [comment]
trigger uts

File: idp-service/src/main/java/io/harness/idp/gitintegration/service/GitIntegrationServiceImpl.java [hunk 1]
```diff
@@ -168,6 +172,21 @@ public void createOrUpdateAppConfigForGitIntegrations(String accountIdentifier,
     if (connectorType == ConnectorType.GITHUB && GitIntegrationUtils.checkIfGithubAppConnector(connectorInfoDTO)) {
       connectorTypeAsString = connectorTypeAsString + SUFFIX_FOR_GITHUB_APP_CONNECTOR;
     }
+
+    if (connectorType == ConnectorType.BITBUCKET
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/gitintegration/service/GitIntegrationServiceImpl.java:176]: nit: we could simplify the if condition. !host.equals(GitIntegrationConstants.HOST_FOR_BITBUCKET_CLOUD) is common

**Review by vikyathharekal — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger ut2

File: idp-service/src/main/java/io/harness/idp/configmanager/utils/ConfigManagerUtils.java [hunk 2]
```diff
@@ -85,8 +92,12 @@ public String getIntegrationConfigBasedOnConnectorType(String connectorType) {
         return readFileFromClassPath(GITLAB_CONFIG_FILE);
       case "AzureRepo":
         return readFileFromClassPath(AZURE_CONFIG_FILE);
-      case "Bitbucket":
-        return readFileFromClassPath(BITBUCKET_CONFIG_FILE);
+      case "Bitbucket_Cloud":
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/configmanager/utils/ConfigManagerUtils.java:95]: nit: underscore looks odd for string value. We can just use - BitbucketCloud, BitbucketServerAuth, BitbucketServerPat

File: idp-service/src/main/java/io/harness/idp/gitintegration/utils/GitIntegrationUtils.java [hunk 3]
```diff
@@ -80,6 +81,12 @@ public boolean checkIfGithubAppConnector(ConnectorInfoDTO connectorInfoDTO) {
         : false;
   }
 
+  public boolean checkIfApiAccessEnabledForBitbucketConnector(ConnectorInfoDTO connectorInfoDTO) {
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/gitintegration/utils/GitIntegrationUtils.java:84]: nit: Code related to specific connector can go into their respective processor class. If we want to write it here as utils, then ideally it should be generic supported.

**Comment by devesh-harness** [comment]
trigger review

**Pull request merged**